File: native/iperf3/src/iperf_error.c
/*
 * Simple iperf error handling for iOS
 */
#include <stdio.h>
#include <errno.h>
#include <string.h>
#include <stdlib.h>
#include <stdarg.h>
#include <pthread.h>
#include "iperf.h"
#include "iperf_api.h"

int gerror = 0;
int i_errno = IENONE;
char iperf_timestrerr[100];

static pthread_mutex_t error_mutex = PTHREAD_MUTEX_INITIALIZER;

/* Do a printf to stderr. */
void
iperf_err(struct iperf_test *test, const char *format, ...)
{
    va_list argp;
    char str[1000];

    pthread_mutex_lock(&error_mutex);

    va_start(argp, format);
    vsnprintf(str, sizeof(str), format, argp);
    va_end(argp);

    if (test != NULL && test->logfile != NULL) {
        fprintf(test->logfile, "%s", str);
        fflush(test->logfile);
    } else {
        fprintf(stderr, "%s", str);
    }

    pthread_mutex_unlock(&error_mutex);
}

/* Do a printf to stderr and exit. */
void
iperf_errexit(struct iperf_test *test, const char *format, ...)
{
    va_list argp;
    char str[1000];

    va_start(argp, format);
    vsnprintf(str, sizeof(str), format, argp);
    va_end(argp);

    if (test != NULL && test->logfile != NULL) {
        fprintf(test->logfile, "%s", str);
        fflush(test->logfile);
    } else {
        fprintf(stderr, "%s", str);
    }

    exit(1);
}

void
iperf_exit(struct iperf_test *test, int status)
{
    exit(status);
}

char *
iperf_strerror(int int_errno)
{
    static char errstr[256];
    int len, perr, herr;

    len = sizeof(errstr);
    memset(errstr, 0, len);

    switch (int_errno) {
    case IENONE:
        snprintf(errstr, len, "no error");
        break;
    case IENEWTEST:
        snprintf(errstr, len, "unable to create a new test");
        perr = 1;
        break;
    case IEINITTEST:
        snprintf(errstr, len, "test initialization failed");
        break;
    case IELISTEN:
        snprintf(errstr, len, "unable to listen for connections");
        perr = 1;
        break;
    case IECONNECT:
        snprintf(errstr, len, "unable to connect to server");
        perr = 1;
        break;
    case IEACCEPT:
        snprintf(errstr, len, "unable to accept connection from client");
        perr = 1;
        break;
    case IESENDCOOKIE:
        snprintf(errstr, len, "unable to send cookie to server");
        perr = 1;
        break;
    case IERECVCOOKIE:
        snprintf(errstr, len, "unable to receive cookie at server");
        perr = 1;
        break;
    case IECTRLWRITE:
        snprintf(errstr, len, "unable to write to the control socket");
        perr = 1;
        break;
    case IECTRLREAD:
        snprintf(errstr, len, "unable to read from the control socket");
        perr = 1;
        break;
    case IECTRLCLOSE:
        snprintf(errstr, len, "control socket has closed unexpectedly");
        break;
    case IEMESSAGE:
        snprintf(errstr, len, "received an unknown message");
        break;
    case IESENDMESSAGE:
        snprintf(errstr, len, "unable to send control message");
        perr = 1;
        break;
    case IERECVMESSAGE:
        snprintf(errstr, len, "unable to receive control message");
        perr = 1;
        break;
    case IESENDPARAMS:
        snprintf(errstr, len, "unable to send parameters to server");
        perr = 1;
        break;
    case IERECVPARAMS:
        snprintf(errstr, len, "unable to receive parameters from client");
        perr = 1;
        break;
    case IEPACKAGERESULTS:
        snprintf(errstr, len, "unable to package results");
        perr = 1;
        break;
    case IESENDRESULTS:
        snprintf(errstr, len, "unable to send results");
        perr = 1;
        break;
    case IERECVRESULTS:
        snprintf(errstr, len, "unable to receive results");
        perr = 1;
        break;
    case IESELECT:
        snprintf(errstr, len, "select failed");
        perr = 1;
        break;
    case IECLIENTTERM:
        snprintf(errstr, len, "the client has terminated");
        break;
    case IESERVERTERM:
        snprintf(errstr, len, "the server has terminated");
        break;
    case IEACCESSDENIED:
        snprintf(errstr, len, "the server is busy running a test. try again later");
        break;
    case IENOSCTP:
        snprintf(errstr, len, "SCTP not supported");
        break;
    case IEUNIMP:
        snprintf(errstr, len, "function not implemented");
        break;
    /* Stream errors (200-series) */
    case IEINITSTREAM:
        snprintf(errstr, len, "unable to initialize stream");
        perr = 1;
        break;
    case IESTREAMLISTEN:
        snprintf(errstr, len, "unable to start stream listener");
        perr = 1;
        break;
    case IESTREAMCONNECT:
        snprintf(errstr, len, "unable to connect stream");
        perr = 1;
        break;
    case IESTREAMACCEPT:
        snprintf(errstr, len, "unable to accept stream connection");
        perr = 1;
        break;
    case IESTREAMWRITE:
        snprintf(errstr, len, "unable to write to stream socket");
        perr = 1;
        break;
    case IESTREAMREAD:
        snprintf(errstr, len, "unable to read from stream (check network/firewall)");
        perr = 1;
        break;
    case IESTREAMCLOSE:
        snprintf(errstr, len, "stream has closed unexpectedly");
        break;
    case IESTREAMID:
        snprintf(errstr, len, "stream has invalid ID");
        break;
    /* Timer errors (300-series) */
    case IENEWTIMER:
        snprintf(errstr, len, "unable to create new timer");
        perr = 1;
        break;
    case IEUPDATETIMER:
        snprintf(errstr, len, "unable to update timer");
        perr = 1;
        break;
    default:
        snprintf(errstr, len, "unknown error (code=%d)", int_errno);
        break;
    }

    return errstr;
}

File: lib/features/speed_test/domain/entities/iperf_error.dart
/// Classification of iperf3 failures, derived from the native `errorCode`
/// (iperf3's internal `i_errno`) returned by the JNI bridge.
///
/// The bridge passes `i_errno` straight through as the `errorCode` field of
/// the run-client result map, plus two sentinels: `-999` for a user cancel
/// and `-1` for a setup failure with no JSON output.
enum IperfErrorCategory {
  /// User cancelled the test (or it was interrupted on purpose).
  cancelled,

  /// Test could not be created/initialized - failed before any traffic.
  setup,

  /// Could not reach the server: control or data socket never connected.
  cannotConnect,

  /// Control channel dropped mid-handshake or mid-test (cookie/params/
  /// results exchange, control socket read/write, select()).
  controlChannelLost,

  /// Data stream was interrupted mid-transfer: stream socket read/write
  /// failed or the peer closed it unexpectedly. This is the "unable to
  /// read stream" case - usually a symptom of the connection going bad
  /// after the test was already running.
  streamInterrupted,

  /// Server is busy running another test.
  serverBusy,

  /// The peer (client or server) terminated the test.
  peerTerminated,

  /// Anything not specifically classified.
  unknown;

  /// Short label for logs / debugging.
  String get debugLabel {
    switch (this) {
      case IperfErrorCategory.cancelled:
        return 'cancelled';
      case IperfErrorCategory.setup:
        return 'setup-failure';
      case IperfErrorCategory.cannotConnect:
        return 'cannot-connect';
      case IperfErrorCategory.controlChannelLost:
        return 'control-channel-lost';
      case IperfErrorCategory.streamInterrupted:
        return 'stream-interrupted';
      case IperfErrorCategory.serverBusy:
        return 'server-busy';
      case IperfErrorCategory.peerTerminated:
        return 'peer-terminated';
      case IperfErrorCategory.unknown:
        return 'unknown';
    }
  }

  /// User-facing message for this failure category.
  String get userMessage {
    switch (this) {
      case IperfErrorCategory.cancelled:
        return 'Test cancelled.';
      case IperfErrorCategory.setup:
        return 'Unable to start the speed test. Please try again.';
      case IperfErrorCategory.cannotConnect:
        return 'Could not reach the test server. '
            'Check your network connection.';
      case IperfErrorCategory.controlChannelLost:
        return 'Lost connection to the test server during setup.';
      case IperfErrorCategory.streamInterrupted:
        return 'The connection dropped during the test. '
            'Check your signal and try again.';
      case IperfErrorCategory.serverBusy:
        return 'The test server is busy. Please try again in a moment.';
      case IperfErrorCategory.peerTerminated:
        return 'The test server ended the test unexpectedly.';
      case IperfErrorCategory.unknown:
        return 'Unable to connect to server. '
            'Please check your internet connection.';
    }
  }
}

/// Map a native iperf3 `errorCode` to a [IperfErrorCategory].
///
/// Codes mirror the `i_errno` enum in `iperf_api.h`. `null` (no code
/// reported) maps to [IperfErrorCategory.unknown].
IperfErrorCategory classifyIperfError(int? code) {
  if (code == null) {
    return IperfErrorCategory.unknown;
  }

  switch (code) {
    // Bridge sentinels.
    case -999:
      return IperfErrorCategory.cancelled;
    case -1:
      return IperfErrorCategory.setup;

    // Test/stream creation & initialization.
    case 100: // IENEWTEST
    case 101: // IEINITTEST
    case 200: // IECREATESTREAM
    case 201: // IEINITSTREAM
      return IperfErrorCategory.setup;

    // Could not establish a connection to the server.
    case 103: // IECONNECT
    case 203: // IESTREAMCONNECT
      return IperfErrorCategory.cannotConnect;

    // Control channel problems (handshake or mid-test control traffic).
    case 105: // IESENDCOOKIE
    case 106: // IERECVCOOKIE
    case 107: // IECTRLWRITE
    case 108: // IECTRLREAD
    case 109: // IECTRLCLOSE
    case 110: // IEMESSAGE
    case 111: // IESENDMESSAGE
    case 112: // IERECVMESSAGE
    case 113: // IESENDPARAMS
    case 114: // IERECVPARAMS
    case 116: // IESENDRESULTS
    case 117: // IERECVRESULTS
    case 118: // IESELECT
      return IperfErrorCategory.controlChannelLost;

    // Peer terminated the test.
    case 119: // IECLIENTTERM
    case 120: // IESERVERTERM
      return IperfErrorCategory.peerTerminated;

    // Server busy with another test.
    case 121: // IEACCESSDENIED
      return IperfErrorCategory.serverBusy;

    // Data stream interrupted mid-transfer.
    case 205: // IESTREAMWRITE
    case 206: // IESTREAMREAD
    case 207: // IESTREAMCLOSE
      return IperfErrorCategory.streamInterrupted;

    default:
      return IperfErrorCategory.unknown;
  }
}
